edlib/include/edlib.hpp

```diff
@@ -222,8 +224,9 @@ extern "C" {
      * @return  Result of alignment, which can contain edit distance, start and end locations and alignment path.
      *          Make sure to clean up the object using edlibFreeAlignResult() or by manually freeing needed members.
      */
-    EdlibAlignResult edlibAlign(const char* query, int queryLength,
-                                const char* target, int targetLength,
+    template<class Element, class AlphabetIdx=uint32_t>
```

Comment:
There is a problem if we set AlphabetIdx to `unsigned char`. That way if the users change Element to anything else (for example int) then they also have to change AlphabetIdx. Changing AlphabetIdx by users does not make much sense because indexing is just an implementation detail. I think setting Element by default to `char` and AlphabetIdx to `uint16_t` is better. Anyway, it's up to you.

Reply:
I see what you mean.
Let's try to break this down. We are transforming sequences from whatever user makes them, into unsigned int of some size. 

First question: why are we even doing this? Why are we doing this transformation? I have to admit, I can't actually exactly remember what was the reason to introduce that, but it has a lot to do with Peq table I believe. Ok, while this is something we could analyze and maybe even consider removing this transformation step, I would rather not get into it now, since answer will not be trivial. Ok, so let's forget this question :D.

Second question then: how do we pick the default size of unsigned int? If we take bigger size, edlib will be consuming more memory. If we pick smaller size, it is more likely that this default value will not be a fit for user and they will have to modify it.
So it all comes down to understanding use cases. We intuitively narrowed down to uint8 (which supports alphabets up to size of 256) and uint16 (which is up to 65536 alphabet size).
When we look at the use cases:
  - genomics -> at many use cases alphabets are pretty small (ACTG...)
  - natural text -> alphabets are in most cases pretty small, smaller than 256.
And I don't know of many more use cases hm. There is one of Paolo, but that seems pretty rare.
So if we go with int8, we probably took care of let's say 90% use cases, and we used the least memory for them. For the 10%, they will need to specify uint16.
If we go with uint16, we probably took care of 99% use cases, but for 90% of them, edlib is consuming double the memory than it could be consuming.

So the question is: would we rather have 10% people have to specify uint16, or would we rather have 90% of people consume double the memory? I don't know how important memory is though.
Aaaaah hard choice :)! I still feel, at the end of this analysis, that we should go with uint8 though -> I just feel very bugged by us consuming unneeded memory.
Let me know how this sounds. We can always bump it up to uint16 at some later moment, if we figure out that is better way to do it.

Reply:
Nice analysis! Let's go with uint8 to figure it out later.